File: templateMaker/src/getMassWeights.cpp
#include "getMassWeights.hpp"

RNode getMassWeights::run(RNode d)
{

  auto BreitWigner = [](float Q2, float M2, float G2) -> float {
    return 1. / ((Q2 - M2) * (Q2 - M2) + M2 * G2);
  };

  auto getBWVec = [&](float Q) {
    ROOT::VecOps::RVec<float> v;
    v.emplace_back(1.);
    float w1 = BreitWigner(Q * Q, 80.419002 * 80.419002, 2.0476 * 2.0476) /
         BreitWigner(Q * Q, 80.319002 * 80.319002, 2.0476 * 2.0476);
    v.emplace_back(w1);
    float w2 = BreitWigner(Q * Q, 80.419002 * 80.419002, 2.0476 * 2.0476) /
         BreitWigner(Q * Q, 80.519002 * 80.519002, 2.0476 * 2.0476);
    v.emplace_back(w2);
    
    return v;
  };

  auto d1 = d.Define("massWeights", getBWVec, {"Wmass_preFSR"});

  //define mass weights as variations
  std::vector<std::string> masses = {"", "massUp", "massDown"};
  getMassWeights::vary("massWeights", true, masses);
  return d1;
}

File: templateMaker/python/getMassWeights.py
from module import *
import h5py
import numpy as np
import matplotlib.pyplot as plt
import mplhep as hep
from math import pi, sqrt

class getMassWeights(module):
   
    def __init__(self, syst = ""):
        pass
      
    def run(self,d):

        @ROOT.Numba.Declare(["RVec<float>"], "RVec<float>")
        def getMassWeights(fullvec): # corresponding to a variation of 50 MeV
            red = np.zeros((2,),dtype='float32')
            red[0]=fullvec[16]
            red[1]=fullvec[6]
            return red
        
        @ROOT.Numba.Declare(["RVec<float>", "RVec<float>"], "RVec<float>")
        def vecMultiplication (v1, v2):
            products = np.zeros((v1.shape[0],v2.shape[0]),dtype='float32')
            for i in range(products.shape[0]):
              for j in range(products.shape[1]):
                products[i,j] = v1[i]*v2[j]
            return products.ravel()

        self.d = d
        self.d = self.d.Define("massWeights","Numba::getMassWeights(LHEReweightingWeightCorrectMass)").Define("helmassWeights", "Numba::vecMultiplication(helWeights,massWeights)")\
            .Define("nhelmassWeights", "helmassWeights.size()")
        
        return self.d

    def getTH1(self):

        return self.myTH1

    def getTH2(self):

        return self.myTH2  

    def getTH3(self):

        return self.myTH3

    def getTHN(self):

        return self.myTHN

    def getGroupTH1(self):

        return self.myTH1Group

    def getGroupTH2(self):

        return self.myTH2Group  

    def getGroupTH3(self):

        return self.myTH3Group  

    def getGroupTHN(self):

        return self.myTHNGroup

    def reset(self):

        self.myTH1 = []
        self.myTH2 = []
        self.myTH3 = [] 
        self.myTHN = [] 

        self.myTH1Group = []
        self.myTH2Group = []
        self.myTH3Group = [] 
        self.myTHNGroup = [] 
